lib/teleterm/services/connectmycomputer/connectmycomputer.go

```diff
@@ -324,6 +330,119 @@ type Provisioner interface {
 	DeleteToken(ctx context.Context, token string) error
 }
 
+type NodeJoinWait struct {
+	cfg *NodeJoinWaitConfig
+}
+
+func NewNodeJoinWait(cfg *NodeJoinWaitConfig) (*NodeJoinWait, error) {
+	err := cfg.CheckAndSetDefaults()
+	if err != nil {
+		return nil, err
+	}
+
+	return &NodeJoinWait{cfg: cfg}, nil
+}
+
+// Run grabs the host UUID of an agent from disk and then waits for the node with the given name to
+// show up in the cluster.
+//
+// The Electron app calls this method soon after starting the agent process.
+func (n *NodeJoinWait) Run(ctx context.Context, accessAndIdentity AccessAndIdentity, cluster *clusters.Cluster) (clusters.Server, error) {
+	nodeName, err := n.getNodeNameFromHostUUIDFile(ctx, cluster)
+	if err != nil {
+		return clusters.Server{}, err
+	}
+
+	server, err := n.waitForNode(ctx, accessAndIdentity, cluster, nodeName)
+	if err != nil {
+		return clusters.Server{}, trace.Wrap(err)
+	}
+
+	return clusters.Server{
+		URI:    cluster.URI.AppendServer(server.GetName()),
+		Server: server,
+	}, nil
+}
+
+func (n *NodeJoinWait) getNodeNameFromHostUUIDFile(ctx context.Context, cluster *clusters.Cluster) (string, error) {
+	hostUUIDPath := filepath.Join(n.cfg.AgentsDir, cluster.ProfileName, "data", utils.HostUUIDFile)
+
+	// NodeJoinWait gets executed when the agent is booting up, so the host UUID file might not exist
+	// on disk yet. Use a ticker to periodically check for its existence.
+	ticker := time.NewTicker(50 * time.Millisecond)
+	defer ticker.Stop()
+
+	for {
```

Comment:
Should we have some sort of timeout here so we don't loop endlessly in the case that there's a problem creating the file?

Reply:
To address both of your comments about timeouts, I didn't add any in the Go code because I expect the client to set for the whole request. The frontend app is going to abort this request after 20 seconds.

However, as you said, both operations essentially block forever if an error happens somewhere else. I suppose a timeout is warranted after all because of that, so I'm going to add a 1 minute timeout to `NodeJoinWait.Run`.